[SPARK-19018][SQL] Add support for custom encoding on csv writer

crafty-coder opened the pull request:
## What changes were proposed in this pull request?

Add support for custom encoding on csv writer, see https://issues.apache.org/jira/browse/SPARK-19018

## How was this patch tested?

Added two unit tests in CSVSuite

crafty-coder committed:
[SPARK-19018][SQL] Add support for custom encoding on csv writer

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented:
ok to test

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:
```diff
@@ -513,6 +513,43 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
+        val df = spark
+          .read
+          .option("header", "false")
+          .option("encoding", encoding)
```
I think our CSV read encoding option is incomplete for now .. there are many discussions about this now. I am going to fix the read path soon. Let me revisit this after fixing it.

HyukjinKwon replied:
Now it's fine. I think we decided to support encoding in CSV/JSON datasources. Ignore the comment above. We can proceed separately.

SparkQA commented:
**[Test build #88779 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88779/testReport)** for PR 20949 at commit [`b9a7bf0`](https://github.com/apache/spark/commit/b9a7bf03b312da151e1d7e37338092bbf5bcb38a).

SparkQA commented:
**[Test build #88779 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88779/testReport)** for PR 20949 at commit [`b9a7bf0`](https://github.com/apache/spark/commit/b9a7bf03b312da151e1d7e37338092bbf5bcb38a).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88779/
Test FAILed.

crafty-coder committed:
Merge remote-tracking branch 'upstream/master'

SparkQA commented:
**[Test build #90274 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90274/testReport)** for PR 20949 at commit [`8fd3e0e`](https://github.com/apache/spark/commit/8fd3e0e40c53fa3674ed1fee6adcac237b8cabc1).

SparkQA commented:
**[Test build #90274 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90274/testReport)** for PR 20949 at commit [`8fd3e0e`](https://github.com/apache/spark/commit/8fd3e0e40c53fa3674ed1fee6adcac237b8cabc1).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90274/
Test FAILed.

crafty-coder committed:
Fix checkstyle issue

SparkQA commented:
**[Test build #90275 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90275/testReport)** for PR 20949 at commit [`0d0addf`](https://github.com/apache/spark/commit/0d0addfd45dac4dc5fe4b32a0a2097990ba24cb5).

SparkQA commented:
**[Test build #90275 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90275/testReport)** for PR 20949 at commit [`0d0addf`](https://github.com/apache/spark/commit/0d0addfd45dac4dc5fe4b32a0a2097990ba24cb5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90275/
Test FAILed.

rberenguel commented:
I've been giving a look to this PR (I've hit this problem in the past and had a chat with @crafty-coder about it and his fixes, too), is there anything we could do to move it forward?

Also, is there any way to trigger a rebuild on Jenkins without adding a dummy commit? Looks like the JVM on this test run blew the heap, just a re-run should be enough (cc @holdenk @HyukjinKwon )

crafty-coder commented:
I would say this change has value on its own. 

At the moment the csv reader applies the charset config but the csv writer is ignoring it, which I think its a bit confusing.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #90279 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90279/testReport)** for PR 20949 at commit [`0d0addf`](https://github.com/apache/spark/commit/0d0addfd45dac4dc5fe4b32a0a2097990ba24cb5).

SparkQA commented:
**[Test build #90279 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90279/testReport)** for PR 20949 at commit [`0d0addf`](https://github.com/apache/spark/commit/0d0addfd45dac4dc5fe4b32a0a2097990ba24cb5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90279/
Test PASSed.

AmplabJenkins commented:
Can one of the admins verify this patch?

gatorsmile commented:
cc @MaxGekk @HyukjinKwon

MaxGekk commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:
```diff
@@ -512,6 +512,43 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils with Te
   }
 
+  test("Save csv with custom charset") {
+    Seq("iso-8859-1", "utf-8", "windows-1250").foreach { encoding =>
```
Could you check the `UTF-16` and `UTF-32` encoding too. The written csv files must contain [BOMs](https://en.wikipedia.org/wiki/Byte_order_mark) for such encodings. I am not sure that Spark CSV datasource is able to read it in per-line mode (`multiLine` is set to `false`). Probably, you need to switch to multLine mode or read the files by Scala's library like in JsonSuite: https://github.com/apache/spark/blob/c7e2742f9bce2fcb7c717df80761939272beff54/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala#L2322-L2338

MaxGekk commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:
```diff
@@ -512,6 +512,43 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils with Te
+        val originalDF = Seq("µß áâä ÁÂÄ").toDF("_c0")
+        // scalastyle:on
+        originalDF.write
+          .option("header", "false")
```
The header flag is disabled by default. Just in case, are there any specific reasons fro testing without CSV header?

crafty-coder replied:
My bad, there is no reason. It's fixed on the next commit.

MaxGekk commented on python/pyspark/sql/readwriter.py:
```diff
@@ -895,6 +895,8 @@ def csv(self, path, mode=None, compression=None, sep=None, quote=None, escape=No
                                           the quote character. If None is set, the default value is
                                           escape character when escape and quote characters are
                                           different, ``\0`` otherwise..
+        :param encoding: sets encoding used for encoding the file. If None is set, it
```
Could you reformulate this `encoding used for encoding`

MaxGekk commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVFileFormat.scala:
```diff
@@ -146,7 +148,13 @@ private[csv] class CsvOutputWriter(
+    context,
+    new Path(path),
+    charset
+  )
```
Move the `)` up like `charset)`. See https://github.com/databricks/scala-style-guide

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:
```diff
@@ -512,6 +512,43 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils with Te
+    Seq("iso-8859-1", "utf-8", "windows-1250").foreach { encoding =>
+      withTempDir { dir =>
+        val csvDir = new File(dir, "csv").getCanonicalPath
+        // scalastyle:off
```
Let's ignore the specific rule for this, e.g.:

```
// scalastyle:off nonascii
...
// scalastyle:on nonascii
```

HyukjinKwon commented:
ok to test

SparkQA commented:
**[Test build #92277 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92277/testReport)** for PR 20949 at commit [`0d0addf`](https://github.com/apache/spark/commit/0d0addfd45dac4dc5fe4b32a0a2097990ba24cb5).

SparkQA commented:
**[Test build #92277 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92277/testReport)** for PR 20949 at commit [`0d0addf`](https://github.com/apache/spark/commit/0d0addfd45dac4dc5fe4b32a0a2097990ba24cb5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92277/
Test PASSed.

crafty-coder committed:
Add test to check UTF-16 and UTF-32

SparkQA commented:
**[Test build #93113 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93113/testReport)** for PR 20949 at commit [`91f4750`](https://github.com/apache/spark/commit/91f4750ff2f4781cea2fb23b1339659abf65009a).

SparkQA commented:
**[Test build #93113 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93113/testReport)** for PR 20949 at commit [`91f4750`](https://github.com/apache/spark/commit/91f4750ff2f4781cea2fb23b1339659abf65009a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

MaxGekk commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:
```diff
@@ -512,6 +513,44 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils with Te
     }
   }
 
+  test("Save csv with custom charset") {
```
Could you prepend `SPARK-19018` to the test title.

crafty-coder committed:
Add jira ticked (SPARK-19018:) as prefix on the related tests

SparkQA commented:
**[Test build #93164 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93164/testReport)** for PR 20949 at commit [`349e132`](https://github.com/apache/spark/commit/349e132770f3a61e453841c82ddc06d34fe51649).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93164/
Test PASSed.

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:
```diff
@@ -512,6 +513,43 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils with Te
+
+        csvDir.listFiles().filter(_.getName.endsWith("csv")).foreach({ csvFile =>
+          val readback = Files.readAllBytes(csvFile.toPath)
+          val expected = (content + "\n").getBytes(Charset.forName(encoding))
```
Currently, the newline is dependent on Univocity. This test is going to be broken on Windows. Let's use platform's newline

crafty-coder replied:
Good Point!

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:
```diff
@@ -512,6 +513,43 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils with Te
+
+  test("SPARK-19018: error handling for unsupported charsets") {
+    val exception = intercept[SparkException] {
+      withTempDir { dir =>
```
`withTempPath`

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:
```diff
@@ -512,6 +513,43 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils with Te
+    // scalastyle:on nonascii
+
+    Seq("iso-8859-1", "utf-8", "utf-16", "utf-32", "windows-1250").foreach { encoding =>
+      withTempDir { dir =>
```
`withTempDir` -> `withTempPath`

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:
```diff
@@ -512,6 +513,43 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils with Te
+      withTempDir { dir =>
+        val csvDir = new File(dir, "csv")
+
+        val originalDF = Seq(content).toDF("_c0").repartition(1)
```
`toDF("_c0")` -> `toDF()`

HyukjinKwon commented on python/pyspark/sql/readwriter.py:
```diff
@@ -895,6 +895,8 @@ def csv(self, path, mode=None, compression=None, sep=None, quote=None, escape=No
                                           escape character when escape and quote characters are
                                           different, ``\0`` otherwise..
+        :param encoding: sets the encoding (charset) to be used on the csv file. If None is set, it
+                                          uses the default value, ``UTF-8``.
```
Likewise, let's match the doc to JSON's.

crafty-coder committed:
Add styling improvements

- Improve method documentation
- Inline method calls that are not too big
- Use platform newline instead of hardcoded one.
- Replace withTempDir with withTempPath

SparkQA commented:
**[Test build #93226 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93226/testReport)** for PR 20949 at commit [`fd857b0`](https://github.com/apache/spark/commit/fd857b005abba233eb7409479436c0abe4e23e4f).

crafty-coder commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:
```diff
@@ -512,6 +513,43 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils with Te
+          .option("encoding", encoding)
+          .csv(csvDir.getCanonicalPath)
+
+        csvDir.listFiles().filter(_.getName.endsWith("csv")).foreach({ csvFile =>
```
What do you mean?

crafty-coder committed:
Merge branch 'master' into master

SparkQA commented:
**[Test build #93227 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93227/testReport)** for PR 20949 at commit [`b6311e3`](https://github.com/apache/spark/commit/b6311e365c0d8bd6d180dbecfb67302f25e5b620).

SparkQA commented:
**[Test build #93227 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93227/testReport)** for PR 20949 at commit [`b6311e3`](https://github.com/apache/spark/commit/b6311e365c0d8bd6d180dbecfb67302f25e5b620).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93227/
Test FAILed.

crafty-coder committed:
Fix import order on CSVSuite

SparkQA commented:
**[Test build #93229 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93229/testReport)** for PR 20949 at commit [`025958a`](https://github.com/apache/spark/commit/025958a7d9e8a741875db2af8878f60cb07409d3).

SparkQA commented:
**[Test build #93226 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93226/testReport)** for PR 20949 at commit [`fd857b0`](https://github.com/apache/spark/commit/fd857b005abba233eb7409479436c0abe4e23e4f).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93226/
Test FAILed.

SparkQA commented:
**[Test build #93229 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93229/testReport)** for PR 20949 at commit [`025958a`](https://github.com/apache/spark/commit/025958a7d9e8a741875db2af8878f60cb07409d3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93229/
Test PASSed.

HyukjinKwon reviewed (approved):
LGTM too

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #93531 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93531/testReport)** for PR 20949 at commit [`025958a`](https://github.com/apache/spark/commit/025958a7d9e8a741875db2af8878f60cb07409d3).

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:
```diff
@@ -514,6 +516,41 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils with Te
+          .option("encoding", encoding)
+          .csv(csvDir.getCanonicalPath)
+
+        csvDir.listFiles().filter(_.getName.endsWith("csv")).foreach({ csvFile =>
```
nit: `.foreach({` -> `.foreach {` per https://github.com/databricks/scala-style-guide#anonymous-methods

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:
```diff
@@ -514,6 +516,41 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils with Te
+        val csvDir = new File(path, "csv").getCanonicalPath
+        Seq("a,A,c,A,b,B").toDF().write
+          .option("encoding", "1-9588-osi")
+          .csv(csvDir)
```
nit: you could use directly `path.getCanonicalPath`

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:
```diff
@@ -514,6 +516,41 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils with Te
+    Seq("iso-8859-1", "utf-8", "utf-16", "utf-32", "windows-1250").foreach { encoding =>
+      withTempPath { path =>
+        val csvDir = new File(path, "csv")
+        Seq(content).toDF().write
```
nit: `.write.repartition(1)` to make sure we write only one file

SparkQA commented:
**[Test build #93531 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93531/testReport)** for PR 20949 at commit [`025958a`](https://github.com/apache/spark/commit/025958a7d9e8a741875db2af8878f60cb07409d3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

HyukjinKwon commented:
Merged to master.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93531/
Test PASSed.

HyukjinKwon commented:
@crafty-coder, what's your JIRA ID? I should know it to assign the JIRA to you.

crafty-coder commented:
@HyukjinKwon and @MaxGekk thanks for your help in this PR!

My JIRA Id is also **crafty-coder**
